### ISSUE TITLE
KOKKOS-Cuda SW not conserving energy

### ISSUE THREAD

**Issue by andeplane** [lammps/lammps#241]
Hi,

I'm running the Cuda compiled KOKKOS testing the [Stillinger-Weber benchmark](http://lammps.sandia.gov/bench/bench_sw.tar.gz). I'm using freshly installed Ubuntu 16.04 LTS, Cuda 8.0 and unmodified LAMMPS installation (both last stable and latest git version). LAMMPS is compiled with _kokkos_cuda_openmpi_. I'm using 2x GTX Titan.

**Edit: I see the same results for LJ benchmark. Energy is not conserved there either, and loses atoms after 30 timesteps.**

When I run 10 timesteps (thermo 1) of the standard SW, I get:

```
Step Temp E_pair E_mol TotEng Press
Step Temp E_pair E_mol TotEng Press
       0         1000    -138771.2            0      -134635    6866.6499
       1    993.93657   -138746.07            0   -134634.95    6881.2446
       2    975.96204   -138671.55            0   -134634.78    6924.3541
       3    946.71794   -138550.32            0   -134634.51    6993.9369
       4    907.24992   -138386.71            0   -134634.15    7086.6376
       5    858.96877   -138186.56            0    -134633.7    7197.8839
       6    803.59684   -137957.02            0   -134633.19      7322.03
       7    743.10127   -137706.24            0   -134632.63    7452.5488
       8    679.61627   -137443.08            0   -134632.06    7582.2729
       9    615.35692   -137176.72            0   -134631.48    7703.6646
      10    552.52801    -136916.3            0   -134630.93     7809.069
      11    493.23208   -136670.54            0   -134630.43    7890.9578
      12     439.3808   -136447.37            0   -134630.01     7942.257
      13    392.61461    -136253.6            0   -134629.67    7956.8109
      14    354.23489   -136094.61            0   -134629.43    7929.9027
      15    325.15293    -135974.2            0    -134629.3    7858.6852
      16    305.85889   -135894.38            0   -134629.29    7742.4203
      17    296.41254    -135855.4            0   -134629.38    7582.4997
      18     296.4559   -135855.77            0   -134629.57    7382.2798
      19    305.24624   -135892.41            0   -134629.85     7146.788
      20    321.70647   -135960.84            0    -134630.2    6882.3567
Loop time of 2.63318 on 1 procs for 20 steps with 32000 atoms
```

a reasonable output. With KOKKOS Cuda version, I get

```
Step Temp E_pair E_mol TotEng Press
       0         1000    -138771.2            0      -134635    6866.6499
       1         1000   -138746.07            0   -134609.87     6904.556
       2         1000   -138671.55            0   -134535.36    7107.0371
       3         1000   -138551.69            0    -134415.5    7445.0149
       4         1000   -138395.72            0   -134259.52    7759.1427
       5         1000    -138219.5            0    -134083.3    7771.3216
       6         1000   -138045.79            0    -133909.6    7121.5929
       7         1000   -137902.54            0   -133766.34    5440.4216
       8         1000   -137818.44            0   -133682.24    2458.1913
       9         1000   -137815.42            0   -133679.22   -1865.9773
      10         1000   -137898.04            0   -133761.85   -7248.4505
      11         1000   -138041.07            0   -133904.87   -13117.836
      12         1000   -138177.58            0   -134041.39   -18935.715
      13         1000   -138191.55            0   -134055.35   -420183.44
      14         1000   -137924.59            0    -133788.4   -427213.48
      15         1000   -137190.36            0   -133054.16   -448126.33
      16         1000    -135799.7            0    -131663.5   -492836.22
      17         1000    -133621.2            0      -129485   -575247.95
      18         1000   -130691.67            0   -126555.47   -810536.14
      19         1000   -127139.08            0   -123002.89   -1016778.7
      20         1000   -123218.35            0   -119082.16   -1306926.9
Loop time of 0.314422 on 1 procs for 20 steps with 32000 atoms
```

and for longer times, LAMMPS crashes with lost atoms. Note that for KOKKOS OpenMP, everything is normal.

**Comment by akohlmey** [comment]
this sounds like there may be an issue with your compiler toolchain. i cannot reproduce this with gcc 4.9.3 and cuda 7.5 on a GeForce GTX TITAN with compute 3.5

**Comment by akohlmey** [comment]
i suggest you switch compiler and cuda toolkit individually to see, which one is triggering the issue for you.

**Comment by stanmoore1** [comment]
I tried this using CUDA 8.0 and TITAN X and it seems to work fine for me (see below). I did use a slightly older version of LAMMPS: 7 Sep 2016-ICMS. Can you list your compiler version and the mpiexec command with Kokkos options?

Stan

Memory usage per processor = 8.39664 Mbytes
Step Temp E_pair E_mol TotEng Press
       0         1000    -138771.2            0      -134635    6866.6499
       1    993.93657   -138746.07            0   -134634.95    6881.2446
       2    975.96204   -138671.55            0   -134634.78    6924.3541
       3    946.71794   -138550.32            0   -134634.51    6993.9369
       4    907.24992   -138386.71            0   -134634.15    7086.6376
       5    858.96877   -138186.56            0    -134633.7    7197.8839
       6    803.59684   -137957.02            0   -134633.19      7322.03
       7    743.10127   -137706.24            0   -134632.63    7452.5488
       8    679.61627   -137443.08            0   -134632.06    7582.2729
       9    615.35692   -137176.72            0   -134631.48    7703.6646
      10    552.52801    -136916.3            0   -134630.93     7809.069
      11    493.23208   -136670.54            0   -134630.43    7890.9578
      12     439.3808   -136447.37            0   -134630.01     7942.257
      13    392.61461    -136253.6            0   -134629.67    7956.8109
      14    354.23489   -136094.61            0   -134629.43    7929.9027
      15    325.15293    -135974.2            0    -134629.3    7858.6852
      16    305.85889   -135894.38            0   -134629.29    7742.4203
      17    296.41254    -135855.4            0   -134629.38    7582.4997
      18     296.4559   -135855.77            0   -134629.57    7382.2798
      19    305.24624   -135892.41            0   -134629.85     7146.788
      20    321.70647   -135960.84            0    -134630.2    6882.3567

**Comment by andeplane** [comment]
Ok, I tried a full reinstall today and this is what I did to reproduce the error:

Install Ubuntu 16.04 LTS.
Install OpenMPI from apt.
Install CUDA 8.0.44 (newest version from NVIDIA web pages) and add to path (had to disable nouveau drivers).
Download LAMMPS stable version (30Jul16).

Go into lammps directory/src
`make no-all`
`make yes-kokkos`
`make kokkos_cuda_openmpi`

I then run the [script](https://gist.github.com/anonymous/f9858b1f25f86ec78511ee719997076c) by running i.e.
`mpirun -n 1 /path/to/lammps/src/lmp_kokkos_cuda_openmpi -k on -g 1 -in in.lj.kokkos`

The compiler is version 5.3.1. I have also tried with gcc 4.9.3, same problem.

**Comment by stanmoore1** [comment]
Did you change the KOKKOS_ARCH variable in the Makefile to be Pascal61?

**Comment by andeplane** [comment]
All LAMMPS files are unchanged. The only modification I did was `make no-all`and `make yes-kokkos`. The architecture is Kepler35 which is correct for my GTX Titan I think?

**Comment by stanmoore1** [comment]
Yes I was confused, sorry. It looks like you are missing a suffix command "-sf kk"

**Comment by andeplane** [comment]
The suffix is in the [LJ script](https://gist.github.com/anonymous/f9858b1f25f86ec78511ee719997076c).

**Comment by stanmoore1** [comment]
No that should work. For a sanity check of your programming environment and to narrow down the problem can you run something else other than Kokkos on the GPU like the GPU package or another CUDA program?

**Comment by andeplane** [comment]
Yeah I'm starting to wonder if there's something with my cards or something. I will try the GPU package in a few hours.

**Comment by stanmoore1** [comment]
Just to be sure I tried adding "suffix kk" to the input script instead of "-sf kk" on the command line and I also see lost atoms for /bench/in.lj. I have no idea why, I'm looking into this now. Can you try "-sf kk" on the command line instead--I think that could fix your problems.

**Comment by stanmoore1** [comment]
If the suffix command is in the input script, the /kk suffix isn't being appended to the verlet integrator and it is using a non-Kokkos version of run_style verlet. That is bad and is why atoms are being lost. However, if the suffix command is on the command line, the /kk suffix is appended to the verlet integrator and all is well.

**Comment by andeplane** [comment]
Ahh great news! Why would it then work with OpenMP KOKKOS without -sf kk?

**Comment by stanmoore1** [comment]
I think it is because Kokkos CUDA version of verlet has critically needed memory transfer between host and device that isn't present in the non-Kokkos version of verlet. Removing the memory synchronization leads to undefined behavior. For OpenMP there is only a single memory space so it doesn't matter.

**Comment by stanmoore1** [comment]
The root of the issue is that the default verlet run_style is instantiated before the input script is read, so it is too late to be affected by a suffix command in the input script. If you want to use the suffix command in the input script, you would need to explicitly add a "run_style verlet" command to the input script. 

LAMMPS should error out if a non-Kokkos version of verlet is used with the Kokkos package, so I will add that. I talked to Steve and we decided to make a note on the suffix doc page explaining that run_style won't be affected by the input script suffix command unless it is explicitly included in the input script. 

So bottom line: you will either need to use the command line "-sf kk" or add "run_style verlet" to your input script. Thanks for reporting the issue.

**Comment by stanmoore1** [comment]
LAMMPS will now error out if this happens with the Kokkos package. I still need to update the doc page.

**Comment by stanmoore1** [comment]
I updated the doc page in the SVN repo, should propagate to GIT soon.

**Comment by stanmoore1** [comment]
Fixed in f48b71f46ba66655f1b9a20753737391ba3e42c5
